### ISSUE TITLE
Training process

### ISSUE THREAD

**Issue by maksimallist** [lucidrains/enformer-pytorch#28]
Hello. Can you share the details of neural model training? Did you train it yourself? Did you collect data for training from basenji dataset files? I am unable to reproduce the claimed results during training.

**Comment by fransilvionGenomica** [comment]
Has anyone looked into this yet? I am also interested in this, since training enformer from scratch using your implementation doesn't reproduce same Pearson correlation values (max I am getting is ~0.4).

**Comment by lucidrains** [comment]
@fransilvionGenomica @maksimallist i tried a while ago using TPUs (didn't have access to large cluster of GPU at the time) and didn't hit the mark (got around 0.5-0.6). this was before Ziga officially released their model over at deepmind

the training script i used is all open sourced [here](https://github.com/lucidrains/enformer-tensorflow-sonnet-training-script)

**Comment by lucidrains** [comment]
@fransilvionGenomica are you planning on training it on proprietary data?

**Comment by fransilvionGenomica** [comment]
@lucidrains I am training your pytorch implementation using a single A100 GPU node with the original basenji dataset and gradient accumulation. I was using the following deepmind notebook as the reference: https://github.com/google-deepmind/deepmind-research/blob/master/enformer/enformer-training.ipynb. I do believe that it is possible to train the model on GPUs, since in the recent Borzoi paper from Enformer co-authors they did not use TPUs (https://www.biorxiv.org/content/10.1101/2023.08.30.555582v1). Unfortunately, they don't provide any training script (https://github.com/calico/borzoi).

**Comment by lucidrains** [comment]
@fransilvionGenomica ahh, i have not checked out Borzoi yet, although someone else told me it is the successor to Enformer

why are you still using this repository if Borzoi is the new SOTA?

**Comment by lucidrains** [comment]
@fransilvionGenomica where do you work btw?

**Comment by lucidrains** [comment]
@fransilvionGenomica  it is possible to train on GPU, but you'll be training up to 2 months if it isn't distributed and well optimized. on TPUs, i was doing 3-4 days of training on 64 v3 TPUs

**Comment by fransilvionGenomica** [comment]
Oh I see. It makes sense. Even Borzoi mentioned it took them ~25 days on 2 GPUs. And I am training on a single GPU. I guess, I will just have to wait then. Thanks!

**Comment by lucidrains** [comment]
@fransilvionGenomica that is strange they waited that long. i thought calico had google level resources

**Comment by lucidrains** [comment]
@fransilvionGenomica i'll revisit genomics maybe end of the month. knee deep in other projects at the moment.

**Comment by lucidrains** [comment]
ahh ok, was told that Borzoi is nothing more than Enformer applied to RNA-seq data. ok then using this repository is fine in that case

**Comment by fransilvionGenomica** [comment]
Yes, architecture wise they are very similar. Borzoi is actually less complex.

**Comment by lucidrains** [comment]
@fransilvionGenomica ok, i'll just copy / paste the existing code and remove that complexity for Borzoi later this month after i read the paper. hopefully they got rid of the annoying gamma positions

**Comment by fransilvionGenomica** [comment]
Just curious, have you noticed anything about the batch size while training enformer from scratch? Like, does it have to be relatively big (like at least 32) or can you train decently even if batch size is 1 or 2?

**Comment by lucidrains** [comment]
@fransilvionGenomica it has to be big (32 or 64)

**Comment by lucidrains** [comment]
@fransilvionGenomica the code in this repository isn't even setup for training. i didn't set up synchronized batchnorm, which is required for it to train well.

**Comment by lucidrains** [comment]
@fransilvionGenomica actually let me just throw that in there for now

**Comment by fransilvionGenomica** [comment]
Have you tried to run your enformer implementation with pytorch lightning?

**Comment by lucidrains** [comment]
@fransilvionGenomica no i haven't, as i said above, my training was done in tensorflow sonnet with TPUs, as i had access to a large cluster of TPUs in collaboration with EleutherAI back then

**Comment by lucidrains** [comment]
@fransilvionGenomica if you ever wire up a working training script, always welcome a pull request, in the spirit of open source science.

**Comment by minjaf** [comment]
> @fransilvionGenomica ahh, i have not checked out Borzoi yet, although someone else told me it is the successor to Enformer
> 
> why are you still using this repository if Borzoi is the new SOTA? without reading the paper, did Borzoi set a new SOTA?

What the paper says:
"Performance is difficult to compare directly to Enformer due to differences in data processing"
(https://www.biorxiv.org/content/10.1101/2023.08.30.555582v1.full)

**Comment by lucidrains** [comment]
@minjaf yea, whenever i see the word `competitive` my eyes glaze over; it is just a euphemism for 'close to SOTA but not quite'

**Comment by lucidrains** [comment]
David Kelley's name is on it though, so i guess i need to read it

**Comment by fransilvionGenomica** [comment]
@lucidrains do you have training/validation loss trends left by any chance? for your tensorflow training code I mean.

**Comment by lucidrains** [comment]
@fransilvionGenomica hey yes, actually still have it lying around (thanks wandb) https://api.wandb.ai/links/lucidrains/9ac4x106

**Comment by ZhuJiwei111** [comment]
> @fransilvionGenomica the code in this repository isn't even setup for distributed training. i didn't set up synchronized batchnorm, which is required for it to train well.


hello，may i ask how to fix this? My training time is several times higher when I train with DDP than single GPU (with the same `batch_size`

**Comment by jsomeara** [comment]
what's confusing to me is that apparently you need high batch size to train enformer, yet borzoi was able to be trained on a batch size of just 2.

**Comment by lucidrains** [comment]
@jsomeara really? that sounds way too low

without gradient accumulation?

**Comment by jsomeara** [comment]
@lucidrains Yes, i was surprised too. But according to the paper:

> we reduced their size … such that we could fit them with a batch size of two on either NVIDIA RTX 4090 GPUs or NVIDIA TITAN RTX GPUs.

a later paper says:

> Borzoi contains a total of 186 million parameters and was trained with a batch size of 2 across two A100 GPUs for 25 days (600 hours)…

also see: [https://github.com/calico/borzoi/blob/main/examples/params.json](https://github.com/calico/borzoi/blob/main/examples/params.json)

no grad accum, either. they also used batch norm in their conv blocks too. It makes me wonder if you really do need high batch size to train enformer. Did you notice lower performance / bad convergence when you used smaller batch size while training enformer?

**Comment by jsomeara** [comment]
i think I will run a test to see and post results back here

**Comment by jsomeara** [comment]
it seems that if grad accum is sufficiently high such that effective batch size is 32+, then things will be fine. I previously thought that you might need to make sure that batchnorm sees a bigger batch size, not just smaller micro batch of like 2 for example. I originally thought having small micro batch might be problematic due to how sparse genomic targets can be.
